### PR TITLE
First pass at code quality and structure

### PR DESCRIPTION
**Pull request by Mystifi**
Alright, I was intrigued by the new bot and decided to do some fixer-upping in order to let the bot use more ES6+ features instead of relying so much on older and/or less efficient code. Big changes came from completely refactoring the way the client handles incoming and outgoing messages, refactoring how messages are passed from `Client#parseMessage` to `Client#messageCallback` (to be honest, these could probably either be consolidated or the latter function should be handled in something like `Room#parseMessage`). I've added a few sample commands, reworked how commands are parsed, added a lot more debug code (that should probably be removed later on), et cetera.

This might still be a work in progress, but the bot actually goes online, logs in, and responds to commands without crashing or doing nothing!

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -12,7 +12,7 @@
 		"node": true
 	},
 	"globals": {
-		"Config": false, "debug": false, "toId": false, "stringify": false, "Rooms": false, "Users": false,
+		"Config": false, "debug": false, "toId": false, "Tools": false, "Rooms": false, "Users": false,
```
Comment (HoeenCoder) [inline .eslintrc:15]: forgot to remove debug from globals whoops!

File: parser.js [hunk 2]
```diff
@@ -1,185 +1,132 @@
...
+	case 'deinit':
+		Rooms(roomid).destroy();
+		break;
+	case 'noinit':
+		if (parts[0] === 'joinfailed') {
+			const room = /You are banned from the room "(.*)"/.exec(parts[1]);
+			if (room) debug(`Join failed - Banned from room ${room[1]}`);
+		} else if (parts[1] === 'nonexistent') {
+			const room = /The room "(.*)" does not exist\./.exec(parts[1]);
+			if (room) debug(`Join failed - The room ${room[1]} does not exist or is modjoined`);
```
Comment (HoeenCoder) [inline parser.js]: You can just console.log things like this, I think even non-devs would want to see this kind of info.

File: parser.js [hunk 3]
```diff
@@ -1,185 +1,132 @@
...
-			user.updateGlobalRank(details.group);
-			break;
-		default:
-			debug(`Unhandled message: ${m}`);
+		break;
+	case 'popup':
+		let popup = parts.join('|');
+		if (popup.includes('has banned you from the room')) {
+			const [room, user] = /<p>(.+) has banned you from the room ([^.]+)[.]<\/p><p>To appeal/.exec(popup);
+			debug(`POPUP (ROOMBAN) - Banned from room '${room}' by '${user}'; please inspect the situation`);
```
Comment (HoeenCoder) [inline parser.js]: Can also console.log this.

File: users.js [hunk 4]
```diff
@@ -51,11 +54,10 @@ class User {
...
 
 	/**
-	 * @return {boolean}
+	 * @type {boolean}
 	 */
-	hasDevAccess() {
-		if (['hoeenhero', 'jumbowhales'].indexOf(this.userid) > -1) return true;
-		return false;
+	get isDev() {
+		return ['mystifi', 'hoeenhero', 'jumbowhales'].includes(this.userid);
```
Comment (HoeenCoder) [inline users.js]: You left yourself in the devs array (I assume thats a mistake)

**Review by HoeenCoder — COMMENTED** [review]
Hey thanks for taking the time to do this, its greatly appreciated! The code you wrote looks really solid, I'll be testing it in a few minutes. In the meantime there are a few minor things to address.

File: .eslintrc (hunk 1 above)
Comment (Mystifi) [reply .eslintrc:15]: `debug` is still a global function; if you want, I can consolidate it into `Tools.debug`.

File: parser.js (hunk 2 above)
Comment (Mystifi) [reply parser.js]: :+1:

File: parser.js (hunk 3 above)
Comment (Mystifi) [reply parser.js]: :+1:

File: users.js (hunk 4 above)
Comment (Mystifi) [reply users.js]: Yeah, that was a remnant from testing.

File: .eslintrc (hunk 1 above)
Comment (HoeenCoder) [reply .eslintrc:15]: Was thinking of an older and already removed var, my bad!

**Pull request merged**